opr/grading.go

```diff
@@ -46,18 +48,145 @@ func Avg(list []*OraclePriceRecord) (avg []float64) {
 }
 
 // CalculateGrade takes the averages and grades the individual OPRs
-func CalculateGrade(avg []float64, opr *OraclePriceRecord) float64 {
+func CalculateGrade(avg []float64, opr *OraclePriceRecord, band float64, version uint) float64 {
 	tokens := opr.GetTokens()
 	opr.Grade = 0
 	for i, v := range tokens {
 		if avg[i] > 0 {
-			d := (v.value - avg[i]) / avg[i] // compute the difference from the average
-			opr.Grade = opr.Grade + d*d*d*d  // the grade is the sum of the square of the square of the differences
+			d := (v.Value - avg[i]) / avg[i] // compute the difference from the average
+			if band > 0 {
+				d = ApplyBand(d, band)
+			}
+			opr.Grade = opr.Grade + d*d*d*d // the grade is the sum of the square of the square of the differences
 		}
 	}
 	return opr.Grade
 }
 
+// ApplyBand
+func ApplyBand(diff float64, band float64) float64 {
+	diff = math.Abs(diff)
+	// If the diff is less than the band, then our diff goes to 0
+	if diff <= band {
+		return 0
+	}
+	return diff - band
+}
+
+// GradeMinimum only grades the top 50 honest records. The input must be the records sorted by
+// self reported difficulty.
+func GradeMinimum(sortedList []*OraclePriceRecord, network string, dbht int64) (graded []*OraclePriceRecord) {
+	// No grade algo can handle 0
+	if len(sortedList) == 0 {
+		return nil
+	}
+
+	switch common.OPRVersion(network, dbht) {
+	case 1:
+		return gradeMinimumVersionOne(sortedList)
+	case 2:
+		return gradeMinimumVersionTwo(sortedList)
+	}
+	panic("Grading version unspecified")
+}
+
+// gradeMinimumVersionTwo is version 2 grading algo
+// 1. PoW to top 50
+// 2. Grade with band to top 25
+// 3. Pay top 25 equally (not done here)
+// 4. Grade to 1
+// 5. Wining price is the last one
+func gradeMinimumVersionTwo(sortedList []*OraclePriceRecord) (graded []*OraclePriceRecord) {
+	list := RemoveDuplicateSubmissions(sortedList)
+	if len(list) < 25 {
+		return nil
+	}
+
+	// Find the top 50 with the correct difficulties
+	// 1. top50 is the top 50 PoW
+	top50 := make([]*OraclePriceRecord, 0)
+	for i, opr := range sortedList {
```

Comment:
this needs to iterate over `list` (post duplicate filter) not `sortedList` (pre duplicate filter)